middleware/basic_auth_test.go

```diff
@@ -11,11 +12,139 @@ import (
+			givenConfig:  defaultConfig,
+			whenAuth:     []string{strings.ToUpper(basic) + " " + base64.StdEncoding.EncodeToString([]byte("invalid"))},
+			expectHeader: basic + ` realm=Restricted`,
+			expectErr:    "code=401, message=Unauthorized",
+		},
+		{
+			name:        "nok, not base64 Authorization header",
+			givenConfig: defaultConfig,
+			whenAuth:    []string{strings.ToUpper(basic) + " NOT_BASE64"},
+			expectErr:   "code=400, message=Bad Request, internal=illegal base64 data at input byte 3",
+		},
+		{
+			name:         "nok, missing Authorization header",
+			givenConfig:  defaultConfig,
+			expectHeader: basic + ` realm=Restricted`,
+			expectErr:    "code=401, message=Unauthorized",
+		},
+		{
+			name:         "nok, too many invalid Authorization header",
+			givenConfig:  defaultConfig,
+			whenAuth:     tooManyAuths,
+			expectHeader: basic + ` realm=Restricted`,
+			expectErr:    "code=401, message=Unauthorized",
+		},
+		{
+			name:        "ok, realm",
+			givenConfig: BasicAuthConfig{Validator: validatorFunc, Realm: "someRealm"},
+			whenAuth:    []string{basic + " " + base64.StdEncoding.EncodeToString([]byte("joe:secret"))},
+		},
+		{
+			name:        "ok, realm, case-insensitive header scheme",
+			givenConfig: BasicAuthConfig{Validator: validatorFunc, Realm: "someRealm"},
+			whenAuth:    []string{strings.ToUpper(basic) + " " + base64.StdEncoding.EncodeToString([]byte("joe:secret"))},
+		},
+		{
+			name:         "nok, realm, invalid Authorization header",
+			givenConfig:  BasicAuthConfig{Validator: validatorFunc, Realm: "someRealm"},
+			whenAuth:     []string{strings.ToUpper(basic) + " " + base64.StdEncoding.EncodeToString([]byte("invalid"))},
+			expectHeader: basic + ` realm="someRealm"`,
+			expectErr:    "code=401, message=Unauthorized",
+		},
+		{
+			name:        "nok, validator func returns an error",
+			givenConfig: defaultConfig,
+			whenAuth:    []string{strings.ToUpper(basic) + " " + base64.StdEncoding.EncodeToString([]byte("error:my_error"))},
+			expectErr:   "my_error",
+		},
+		{
+			name: "ok, skipped",
+			givenConfig: BasicAuthConfig{Validator: validatorFunc, Skipper: func(c echo.Context) bool {
+				return true
+			}},
+			whenAuth: []string{strings.ToUpper(basic) + " " + base64.StdEncoding.EncodeToString([]byte("invalid"))},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			e := echo.New()
+
+			mw := BasicAuthWithConfig(tc.givenConfig)
+
+			h := mw(func(c echo.Context) error {
+				return c.String(http.StatusTeapot, "test")
+			})
+
+			req := httptest.NewRequest(http.MethodGet, "/", nil)
+			res := httptest.NewRecorder()
+
+			if len(tc.whenAuth) != 0 {
+				for _, a := range tc.whenAuth {
+					req.Header.Add(echo.HeaderAuthorization, a)
+				}
+			}
+			err := h(e.NewContext(req, res))
+
+			if tc.expectErr != "" {
+				assert.Equal(t, http.StatusOK, res.Code)
+				assert.EqualError(t, err, tc.expectErr)
```

Comment:
Suggestion: you can use tc.wantErr instead of expectedErr and here:
```
assert.Equal(t, http.StatusOK, res.Code)
require.Error(t, err, tc.wantErr)
return
```
so you can get rid of that `else`.

Reply:
after that comes `if tc.expectHeader != "" {` assertion which we want to be executed also